src/Tuura/Concept/FSM/Translation.hs

```diff
@@ -205,6 +210,11 @@ fsmarcxToFsmarc arc = FsmArc newSourceEnc (transx arc) newDestEnc
     where newSourceEnc = (encToInt . srcEncx) arc
           newDestEnc = (encToInt . destEncx) arc
 
+removeInvariants :: [FsmArcX a] -> [[Tristate]] -> [FsmArcX a]
+removeInvariants xs inv = concatMap (\x -> if (destEncx x `elem` inv || srcEncx x `elem` inv)
+                          then []
+                          else [x]) xs
+
 -- Produce all arcs with all X's resolved
-createAllArcs :: Ord a => [([Transition a], Transition a)] -> [FsmArc a]
-createAllArcs = map fsmarcxToFsmarc . expandAllXs . createArcs
+createAllArcs :: Ord a => [([Transition a], Transition a)] -> [[Tristate]] -> [FsmArc a]
+createAllArcs xs inv = map (fsmarcxToFsmarc) (removeInvariants (expandAllXs (createArcs xs)) inv)
```

Comment:
More than likely. I will give it a try, and find it's most simplified form, for less confusion.